[브루니] H2 DB에 회원과 게시글 저장 / EC2 배포

23Yong opened the pull request:
## 주요 작업
* [x] H2 데이터베이스 추가
* [x] 테이블 생성을 위한 `schema.sql` 추가
* [x] 회원가입, 회원 목록 조회, 회원 프로필 조회, 회원 정보 수정 기능 jdbcTemplate을 이용해 DB에 저장하도록 수정
  * `@Transactional 적용`
* [x] 게시글 저장, 게시글 목록 조회, 게시글 상세 조회 기능 jdbcTemplate을 이용해 DB에 저장하도록 수정
  * `@Transactional 적용`
* [x] AWS EC2에 배포

## 학습 키워드
* EC2 (컴퓨팅 리소스, 메모리 제공)
* 배포 스크립트
* `@Transactional`
* 읽기 전용 트랜잭션

## 더 학습해야할 사항
* `@Transactional(readOnly = true)` 옵션을 주었을 때 DB에 읽기 락만을 걸어 성능이 향상 된다고 하는데 읽기 락에 대해 학습해야 할 것 같습니다. 또한 읽기 전용 트랜잭션에서는 예외가 발생할 경우 롤백하지 않습니다. 따라서 예외 처리가 더욱 간단해진다고 합니다.
* JdbcTemplate 을 사용할 때 queryForObject를 사용했는데 해당 메서드는 EmptyDataResultAccessException 를 발생시킵니다. 이를 예외처리하기 위해 try-catch 를 사용했는데 더 나은 방법이 있는지 잘 모르겠습니다.

## 배포 과정
### EC2 인스턴스 생성하기
* Ubuntu Server 20.04 LTS
* t2.micro
* 보안 그룹 설정 (HTTP: 80, HTTPS: 443, SSH: 22, SpringBoot: 8080)
### EC2 서버에 접속
* apt 업데이트
```bash
$ sudo apt-get update && sudo apt-get upgrade
```
* EC2 서버에 접속해 아래와 같은 명령어로 Java11 설치
```bash
$ sudo apt-get install openjdk-11-jdk
```
### 배포
* github에서 코드를 받아올 수 있도록 EC2에 git 설치
```bash
$ sudo apt-get install git
```
* 이후 프로젝트를 저장할 디렉토리 생성 후 git clone으로 프로젝트를 저장
* `./gradlew test` 명령어로 코드들이 수행되는지 검증
* 배포 스크립트 생성 (deploy.sh)
<details>
<summary>배포 스크립트</summary>

```bash
#!/bin/bash

REPOSITORY=/home/ubuntu/app/spring-cafe
PROJECT_NAME=be-java-cafe-max

cd $REPOSITORY/$PROJECT_NAME/

echo "> git pull"

git pull

echo "> spring-cafe 시작"

./gradlew build

echo "> spring-cafe directory로 이동"

cd $REPOSITORY

echo "> Build file copy..."

cp $REPOSITORY/$PROJECT_NAME/build/libs/*.jar $REPOSITORY/

echo "> 현재 구동중인 애플리케이션 pid 확인"

CURRENT_PID=$(pgrep -f cafe-0.0.1-SNAPSHOT.*.jar)

if [ -z "$CURRENT_PID" ]; then
        echo "> 현재 구동 중인 애플리케이션이 없어 종료하지 않습니다."
else
        echo "> kill -15 $CURRENT_PID"
        kill -15 $CURRENT_PID
        sleep 5
fi

echo "> 새 애플리케이션 배포"

JAR_NAME=$(ls -tr $REPOSITORY/ | grep jar | tail -n 1)
echo "> JAR name: $JAR_NAME"

nohup java -jar $REPOSITORY/$JAR_NAME 2>&1 &
```
</details>

* `CURRENT_PID=$(pgrep -f cafe-0.0.1-SNAPSHOT.*.jar)`
  * 기존 수행 중이던 스프링부트 애플리케이션 종료
  * pgrep 은 process id 만 추출하는 명령어
  * -f 옵션은 process 이름으로 탐색하는 옵션
* `nohup java -jar /home/ubuntu/app/spring-cafe/cafe-0.0.1-SNAPSHOT.jar 2>&1 &`
  * nohup : 프로세스가 실행되는 동안에 로그아웃 되어도 프로세스가 종료되지 않도록 하는 명령어
  * java -jar : Java 어플리케이션을 실행하기 위한 명령어
  * 2>&1 : STDERR 출력을 STDOUT으로 보내는 리다이렉션. 이 옵션을 사용하면 프로세스의 표준 오류 출력이 로그에 기록된다.
  * & : 백그라운드에서 프로세스를 실행하도록 하는 명령어

* 스크립트 파일에 실행 권한 추가
```bash
$ chmod +x ./deploy.sh
```
* 스크립트 실행
```
$ ./deploy.sh
```

23Yong committed:
add : H2 데이터베이스 의존성 추가

23Yong committed:
docs : H2 데이터베이스 적용을 위한 datasource 프로퍼티 설정

23Yong committed:
docs : 게시글과 유저정보를 저장하기 위한 테이블 스키마 작성

23Yong committed:
feat : JdbcTemplate을 통해 게시글 생성, 조회 로직을 db와 연결

@Transactional을 통해 하나의 로직이 중간에 실패하더라도 rollback이 될 수 있도록 설정

23Yong committed:
feat : JdbcTemplate을 통해 유저 생성, 조회, 변경 로직을 db와 연결

@Transactional을 통해 하나의 로직이 중간에 실패하더라도 rollback이 될 수 있도록 설정

23Yong committed:
fix : 잘못된 경로 수정

사용자가 쉽게 페이지를 이동할 수 있도록 경로 수정

23Yong committed:
docs : 배포 url 추가

23Yong committed:
fix : 잘못된 fragment 삽입 수정

23Yong committed:
fix : 잘못된 url 경로 수정

23Yong committed:
refactor : DateTimeFormatter를 사용하지 않고 Timestamp를 LocalDateTime 타입으로 변환

23Yong committed:
feat : 서버가 죽지 않도록 모든 예외를 잡는 핸들러 추가

23Yong committed:
refactor : 중복되는 `RowMapper`를 필드로 분리

23Yong committed:
feat : `queryForObject`의 `EmptyDataResultAccessException` 예외처리

23Yong committed:
fix : 잘못된 뷰를 반환하고 있던 것을 error 페이지를 보여주도록 수정

wheejuni commented on src/main/java/kr/codesqaud/cafe/repository/impl/UserMemoryRepository.java:
```diff
@@ -37,4 +34,8 @@ public List<User> findAll() {
 	}
+
+	@Override
+	public void update(User user) {
+	}
```
어라 여긴 아무 액션이 없군요.

23Yong replied:
엇, 제가 실수로 누락했습니다.🥲 바로 수정했습니다!

wheejuni reviewed (changes requested):
배포 스크립트에도, 코드에도 딱히 문제가 없어 보입니다 👍 
다만 로직이 빈 곳이 한 곳 보여서, 이 부분은 채우셨으면 좋겠습니다. 여기까지 마무리하고 머지하는게 좋겠습니다.

23Yong committed:
refactor : [리뷰 사항 반영] update 메서드의 동작 추가

기존 `UserMemoryRepository` 클래스의 `update` 메서드는 아무 동작을 하지 않았는데 기존 user를 대체하는 로직을 작성
